connectors/mediasession/src/main/java/com/theoplayer/android/connector/mediasession/PlaybackStateProvider.kt

```diff
@@ -0,0 +1,251 @@
+            addEventListener(PlayerEventTypes.PLAYING, onPlaying)
+            addEventListener(PlayerEventTypes.PAUSE, onPause)
+            addEventListener(PlayerEventTypes.ERROR, onError)
+            addEventListener(PlayerEventTypes.READYSTATECHANGE, onReadyStateChange)
+            addEventListener(PlayerEventTypes.WAITING, onWaiting)
+            addEventListener(PlayerEventTypes.ENDED, onEnded)
+            addEventListener(PlayerEventTypes.DURATIONCHANGE, onDurationChange)
+        }
+    }
+
+    private fun unregisterListeners() {
+        player?.apply {
+            removeEventListener(PlayerEventTypes.LOADSTART, onLoadStart)
+            removeEventListener(PlayerEventTypes.TIMEUPDATE, onTimeUpdate)
+            removeEventListener(PlayerEventTypes.SOURCECHANGE, onSourceChange)
+            removeEventListener(PlayerEventTypes.LOADEDMETADATA, onLoadedMetadata)
+            removeEventListener(PlayerEventTypes.PLAY, onPlay)
+            removeEventListener(PlayerEventTypes.PLAYING, onPlaying)
+            removeEventListener(PlayerEventTypes.PAUSE, onPause)
+            removeEventListener(PlayerEventTypes.ERROR, onError)
+            removeEventListener(PlayerEventTypes.READYSTATECHANGE, onReadyStateChange)
+            removeEventListener(PlayerEventTypes.WAITING, onWaiting)
+            removeEventListener(PlayerEventTypes.ENDED, onEnded)
+            removeEventListener(PlayerEventTypes.DURATIONCHANGE, onDurationChange)
+        }
+    }
+
+    private val onLoadStart = { _: LoadStartEvent ->
+        updatePlaybackState(PlaybackStateCompat.STATE_BUFFERING)
+    }
+
+    private val onTimeUpdate = { _: TimeUpdateEvent ->
+        invalidatePlaybackState()
+    }
+
+    private val onSourceChange = { _: SourceChangeEvent ->
+        if (player?.source != null) {
+            updatePlaybackState(PlaybackStateCompat.STATE_PAUSED)
+        } else {
+            updatePlaybackState(PlaybackStateCompat.STATE_STOPPED)
+        }
+        connector.setMediaSessionMetadata(player?.source)
+    }
+
+    private val onLoadedMetadata = { _: LoadedMetadataEvent ->
+        connector.invalidateMediaSessionMetadata()
+    }
+
+    private val onPlay = { _: PlayEvent ->
+        if (player!!.readyState.ordinal < ReadyState.HAVE_CURRENT_DATA.ordinal) {
+            updatePlaybackState(PlaybackStateCompat.STATE_BUFFERING)
+        }
+    }
+
+    private val onPlaying = { _: PlayingEvent ->
+        updatePlaybackState(PlaybackStateCompat.STATE_PLAYING)
+    }
+
+    private val onPause = { _: PauseEvent ->
+        updatePlaybackState(PlaybackStateCompat.STATE_PAUSED)
+    }
+
+    private val onError = { _: ErrorEvent ->
+        updatePlaybackState(PlaybackStateCompat.STATE_ERROR)
+    }
+
+    private val onReadyStateChange = { event: ReadyStateChangeEvent ->
+        if (event.readyState.ordinal < ReadyState.HAVE_CURRENT_DATA.ordinal) {
+            updatePlaybackState(PlaybackStateCompat.STATE_BUFFERING)
```

Comment:
There are 3 places from where we are setting the state as buffering (onPlay, onReadyStateChange, onWaiting).
Can't we rely on only one of them? for example wouldn't `onWaiting` make the most sense?